### PR TITLE
Add and use OpenVSX caching proxy

### PR DESCRIPTION
**Pull request by corneliusludmann**
This PR introduces a caching proxy for OpenVSX at https://open-vsx.gitpod.io.

**Here are some parameters that are open for discussion:**
- [ ] With the current implementation, every GET, HEAD, and PUSH call is cashed.
- [ ] Every response is cached for 30 minutes (valid time).
- [ ] A cached response is removed when not accessed for 120 minutes.
- [ ] In case of an error (timeout http_500 http_502 http_503 http_504) a stale response in the cache (of the past 120 minutes) is served.
- [ ] Cache folder has max. size of 10 GB.
- [ ] Header from upstream like Cache-Control are ignored (everything is cached for 30 minutes).
- [ ] For VSCode the OpenVSX URL is changed in the leeway.Dockerfile, Theia uses the TheiaPluginService, right? Or does Theia need a URL change as well?

**How to test:**
- Open a workspace
- Use the extensions view to search for extensions
- Observe the proxy logs: `kubectl logs -f deployment/proxy | grep proxying`, every `"cacheStatus":"HIT"` is a cached result

### CODE REVIEW

**Comment by aledbf** [comment]
@corneliusludmann this should be allowed only from workspaces running in `.gitpod.io` (not an open proxy)

**Comment by corneliusludmann** [comment]
> @corneliusludmann this should be allowed only from workspaces running in `.gitpod.io` (not an open proxy)

@aledbf: Good idea. Do you have a hint for me on how to do this?

**Comment by svenefftinge** [comment]
Also, I remember that @akosyakov mentioned there were CORS issues with accessing open-vsx.org from within frontends loaded from *.gitpod.io. A (public) proxy could solve that issue.

**Comment by akosyakov** [comment]
@svenefftinge CORS issues were sorted out on Open VSX side already: https://github.com/eclipse/openvsx/issues/268#issuecomment-801747761

File: components/ide/code/BUILD.yaml
Comment (csweichel) [inline components/ide/code/BUILD.yaml:16]: if we made this an explicit variable, we could set a workspace default and only produce a build when absolutely necessary. 

```suggestion
        OPENVSX_URL: ${OPENVSX_URL}
```

File: components/server/src/theia-plugin/theia-plugin-service.ts
Comment (csweichel) [inline components/server/src/theia-plugin/theia-plugin-service.ts:43]: This bit I find most interesting. OpenVSX blocking workspace starts is a no-go.
@akosyakov with VSCode resolving its plugins itself, other than for Theia, do we even still need this?

**Comment by csweichel** [comment]
> @corneliusludmann this should be allowed only from workspaces running in `.gitpod.io` (not an open proxy)

Unfortunately this would break self-hosted, unless we introduce all this variance in a "SaaS" only build - or at least produce a VS Code build for Self-Hosted that does not use `openvsx.gitpod.io`.

File: .werft/build.js
Comment (meysholdt) [inline .werft/build.js:110]: Can you make the Domain Name a parameter or env var of the IDE image? 
Introducing it as built tim variance is problematic because the build artefacts should be installable on any domain.

**Review by meysholdt — CHANGES_REQUESTED** [review]

File: .werft/build.js
Comment (corneliusludmann) [reply .werft/build.js:110]: I totally agree. I have made some steps in that direction here: https://github.com/gitpod-io/gitpod/tree/clu/openvsx-proxy-2
However, that doesn't work yet that well.

**Comment by aledbf** [comment]
> Unfortunately this would break self-hosted, 

Why? The default value should be not use a proxy.

> unless we introduce all this variance in a "SaaS" only build - or at least produce a VS Code build for Self-Hosted that does not use `openvsx.gitpod.io`.

`openvsx.gitpod.io` makes sense only for SaaS (unless we want to be in charge of maintaining a self-hosted component running outside the control of gitpod.io)

**Comment by aledbf** [comment]
What's the impact of running an openvsx proxy? (resources like traffic/cpu)

**Comment by corneliusludmann** [comment]
> > Unfortunately this would break self-hosted,
> 
> Why? The default value should be not use a proxy.

When the OpenVSX URL is set during runtime (see https://github.com/gitpod-io/gitpod/pull/3677#discussion_r605601092) a self-hosted installation could use its own proxy for OpenVSX, not https://open-vsx.gitpod.io but https://open-vsx.my-gitpod-installation.com.


> > unless we introduce all this variance in a "SaaS" only build - or at least produce a VS Code build for Self-Hosted that does not use `openvsx.gitpod.io`.
> 
> `openvsx.gitpod.io` makes sense only for SaaS (unless we want to be in charge of maintaining a self-hosted component running outside the control of gitpod.io)

I don't know if it's worth to artificially restrict the access to open-vsx.gitpod.io to our components unless we find an easy way to do this. However, I consider the risk of abuse to be relatively low.

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/3677#" title="Author self-approved">corneliusludmann</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **csweichel** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @csweichel` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [corneliusludmann]
- **[components/ide/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ide/OWNERS)**
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [corneliusludmann]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by svenefftinge** [comment]
I wonder if we can use this approach and configure it so it can bridge longer unavailability of onen-vsx.org.
The data shouldn't be too big (it's only metadata), so we could maybe cache requests for a couple of weeks instead of 30 minutes and only use that cache if the upstream registry doesn't respond in a certain time.

**Comment by csweichel** [comment]
Blobserve can replace strings when serving content to the frontend. Instead of backing in the open-vsx domain at build-time, we should make that domain configurable by virtue of blobserve.

This way we avoid the self-hosted interaction, and can enable a custom OpenVSX proxy in installations where required.

**Comment by csweichel** [comment]
This almost strikes me as new PR, also considering that we're no longer using nginx as proxy.

**Pull request closed without merging**
